[V1][Core] Add worker_base for v1 worker

AoyuQC opened the pull request:
* Creates new WorkerBase class with common configuration
* Modifies GPU Worker to inherit from WorkerBase and keep GPU-specific implementations

comaniac commented on vllm/v1/worker/worker_base.py:
```diff
@@ -0,0 +1,126 @@
+        # Device and model state
+        self.device: Optional[torch.device] = None
+        self.model_runner: Optional[nn.Module] = None
+        self.init_gpu_memory: Optional[int] = None
```
WorkerBase should not be GPU specific?

AoyuQC replied:
removed init_gpu_memory in worker_base

comaniac reviewed (commented):
Also cc @WoosukKwon

youkaichao commented on vllm/v1/worker/worker_base.py:
```diff
@@ -0,0 +1,125 @@
+logger = init_logger(__name__)
+
+
+class WorkerBase(ABC):
```
can we inherit from `vllm.worker.worker_base.WorkerBase` to reduce code duplication, similar to the executor case?

AoyuQC replied:
hi, just inherit from `vllm.worker.worker_base.WorkerBase`. However, it looks wired since we have to do some fake definitions for abstract methods. Like `add_lora`. Please check.

youkaichao replied:
I think you can remove `@abstractmethod` in `vllm.worker.worker_base.WorkerBase` . 

`raise NotImplementedError` there is enough, and then you don't need to add them in v1 worker.

DarkLight1337 commented on vllm/worker/worker_base.py:
```diff
@@ -60,28 +60,24 @@ def init_device(self) -> None:
 
+    @abstractmethod
+    def get_model(self) -> nn.Module:
+        raise NotImplementedError
```
Yes, doing so prevents subclasses from being instantiated if they haven't implemented the abstract method.

youkaichao replied:
> Yes, doing so prevents subclasses from being instantiated if they haven't implemented the abstract method, instead of only erroring out when the method is called.

I think that's too restrictive, and hurts the development. erroring out when the method is called looks enough. we can add the implementation step by step.

AoyuQC replied:
I agree to remove unnecessary `@abstractmethod` and only `raise NotImplementedError`. What do you think?

DarkLight1337 replied:
To avoid forgetting to implement them at the end, can we add a test that prints out which abstract methods remain to be implemented for each worker subclass?

AoyuQC replied:
This confuses me. What's the difference between throwing error when the method is called (`@abstractmethod`) and failing to pass the test for checking abstract methods implementation?

DarkLight1337 replied:
We can make the test print out warnings instead of failing outright.

DarkLight1337 replied:
The difference is that users won't get unnecessary warnings while developers can remain aware of this.

DarkLight1337 replied:
Sounds good to me. You can keep the `abstractmethod` for `get_model` as it's the most straightforward one to implement.

AoyuQC committed:
[V1][Core] Add worker_base for v1 worker

1. reuse WorkerBase in vllm.worker.worker_base
2. remove unnecessary abstract methods and only give warnings for unimplemented methods

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

AoyuQC commented:
In order to make the code easier to read, I've merged it into one commit, and I'm looking forward to your suggestions. @youkaichao and @DarkLight1337

DarkLight1337 commented on vllm/worker/worker_base.py:
```diff
@@ -27,6 +29,64 @@
+    methods.
+    """
+
+    def decorator(cls: Type):
```
```suggestion
    def decorator(cls: Type):
```
Let's use a typevar here so we can keep the original type information.

DarkLight1337 commented on vllm/worker/worker_base.py:
```diff
@@ -27,6 +29,64 @@
+                    base_method_name = base_method.__func__
+                else:
+                    continue
+                class_method = getattr(cls, attr_name, False)
+                # bypass method defined in sub class
+                if not class_method:
+                    continue
+                if class_method == base_method_name:
+                    unimplemented_methods.append(attr_name)
```
```suggestion
                class_method_name = getattr(cls, attr_name, False)
                # bypass method defined in sub class
                if not class_method_name:
                    continue
                if class_method_name == base_method_name:
                    unimplemented_methods.append(attr_name)
```

DarkLight1337 commented on vllm/worker/worker_base.py:
```diff
@@ -27,6 +29,64 @@
+        @wraps(original_init)
+        def wrapped_init(self, *args, **kwargs):
+            original_init(self, *args, **kwargs)
+            unimplemented_methods = []
```
I think we can move this code out into a separate function to improve the readability.

AoyuQC committed:
improve readability by adding warn_unimplemented_methods()

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

DarkLight1337 reviewed (approved):
Thanks for cleaning this up! @youkaichao can you take a final pass at this as well?

AoyuQC committed:
fix for mypy [method-assign] error

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

youkaichao commented on vllm/worker/worker_base.py:
```diff
@@ -26,7 +28,68 @@
+T = TypeVar('T')
 
+
+def check_implementation():
```
I feel this part is too complicated to understand. can we just have `raise NotImplementedError` ?

if we don't get these errors, it means this function is not called right now, and it's fine.

if we get these errors, the developers will definitely see it and implement it, if necessary.

DarkLight1337 replied:
We still `raise NotImplementedError` when the method is used, this code is just to log additional warnings if some methods aren't implemented.

youkaichao replied:
> this code is just to log additional warnings if some methods aren't implemented

why do we need this?

DarkLight1337 replied:
It becomes easier to keep track of the feature completeness as we know which features are still missing without needing someone to try using that feature.

AoyuQC replied:
I think this helps to remind developers that some base class methods aren't implemented in inherited sub classes.

youkaichao commented on vllm/worker/worker_base.py:
```diff
@@ -60,28 +123,26 @@ def init_device(self) -> None:
+    ) -> Optional[List[SamplerOutput]]:
+        raise NotImplementedError
+
+    @avoid_check
```
I feel this is too complicated, we need to decorate `@avoid_check` for every method that has a common implementation in the base class.

DarkLight1337 replied:
We can check `__isabstractmethod__` attribute to only select the methods that are decorated by `@abstractmethod`.

AoyuQC replied:
We probably need to go back to the original point of the discussion which is making development easier, right? Thanks @youkaichao for the reminder, I initially inherited v0 workerbase (`vllm.worker.worker_base.WorkerBase`). However, this base has some `@abstractmethod` that are not needed in v1 right now, such as `add_lora`,`remove_lora`. 

It's easy if you just remove `@abstractmethod`, but it's probably not development-friendly because developers always want a good reminder for what methods aren't implemented right now. Therefore, with the following user experience, there are some unimplemented methods in the v1 GPU.

`WARNING 02-10 12:57:48 worker_base.py:67] Methods add_lora,determine_num_available_blocks,get_cache_block_size_bytes,list_loras,pin_lora,remove_lora not implemented in <vllm.v1.worker.gpu_worker.Worker object at 0x7a025b3297b0>`

As a developer, you can choose to implement it or not, and you don't need to go through trial and error to find a way to achieve it. Right? So the final experience is that we only keep the key abstract methods, and tell the developers the rest through reminders

BTW, @DarkLight1337 , we cannot use `__isabstractmethod__` since we removed most `@abstractmethod`. And there is only one `start_worker_execution_loop` needing @avoid_check, so this shouldn't be a huge amount of work? @youkaichao

DarkLight1337 replied:
How about keeping abstractmethod but effectively disabling them by removing `ABC` base class?

AoyuQC replied:
> How about keeping abstractmethod but effectively disabling them by removing `ABC` base class?

@DarkLight1337 this makes life easier and won't throw errors if we don't implement `@abstractmethod` in base class. **But there was no reminder of this.** If this is an approach you agree with, isn't @youkaichao 's idea a more straightforward approach that doesn't even require an `@abstractmethod`?

DarkLight1337 replied:
I think currently `check_implementation` replaces the functionality of `ABC` (which throws errors instead of warnings if there are unimplemented methods). Basically, we use `abstractmethod` (soft check at initialization) + `raise NotImplementedError` (hard check at call site) to remind ourselves which methods are still missing.

AoyuQC replied:
> I think currently `check_implementation` replaces the functionality of `ABC` (which throws errors instead of warnings if there are unimplemented methods). Basically, we use `abstractmethod` (soft check at initialization) + `raise NotImplementedError` (hard check at call site) to remind ourselves which methods are still missing.

I see. This makes sense to me.

AoyuQC replied:
Hey guys, I really love the open source community's discussion atmosphere, but I also hope this PR doesn't drop halfway. In this case, I don't know if you members have any mechanisms? Like voting or something? No matter what kind of change it is, I might need a clear and uniform suggestion. Thank you! @youkaichao @DarkLight1337

youkaichao commented:
Sorry for the confusion, summary of my reviews, per offline discussion with @DarkLight1337 :

- we should just have `raise NotImplementedError` instead of having both `raise NotImplementedError` and `abstractmethod` (implementation wise, I think you can check the function’s code object to see if it contains `NotImplementedError`)
- we should move `check_implementation`  into `vllm/utils.py` so that it can be used elsewhere as well
- we should rename `check_implementation`  to `warn_for_unimplemented_methods` to be clear

AoyuQC commented:
> Sorry for the confusion, summary of my reviews, per offline discussion with @DarkLight1337 :
> 
>     * we should just have `raise NotImplementedError` instead of having both `raise NotImplementedError` and `abstractmethod` (implementation wise, I think you can check the function’s code object to see if it contains `NotImplementedError`)
> 
>     * we should move `check_implementation`  into `vllm/utils.py` so that it can be used elsewhere as well
> 
>     * we should rename `check_implementation`  to `warn_for_unimplemented_methods` to be clear

@youkaichao Thank you for reply! Some questions:
1. I notice there is `vllm/worker/utils.py`. Do I need to put check_implementation there or still in `vllm/utils.py`?
2. Since we only keep `raise NotImplementedError`, do we still need `class WorkerBase(ABC): `or just `class WorkerBase():` ?

AoyuQC committed:
refactor for easy maintenance

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

AoyuQC committed:
Merge branch 'main' into feature/v1-worker-base

youkaichao commented on vllm/utils.py:
```diff
@@ -2253,3 +2254,61 @@ def import_pynvml():
+    def wrapper(*args: Any, **kwargs: Any) -> T:
+        return func(*args, **kwargs)
+
+    wrapper._avoid_check = True  # type: ignore
```
I think we don't need this `_avoid_check` any more?

AoyuQC replied:
Then you will get warn that method `def start_worker_execution_loop(self) -> None` is not implemented in sub class. Actually it's already implemented in base class`vllm.woker.worker_base.WorkerBase`. Is that acceptable?

youkaichao commented on vllm/worker/worker_base.py:
```diff
@@ -27,7 +28,7 @@
 
 
-class WorkerBase(ABC):
+class WorkerBase:
```
you can just decorate this class

AoyuQC replied:
The reason that I decorate `WorkerBase` in `vllm.v1.worker.worker_base` is that in addition to methods in `vllm.worker.worker_base.WorkerBase`, it will also remind your methods there (`vllm.v1.worker.worker_base.WorkerBase`)

AoyuQC committed:
remove _avoid_check

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

AoyuQC committed:
fix per-commit

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

AoyuQC committed:
remove avoid_check

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

youkaichao commented on vllm/utils.py:
```diff
@@ -2253,3 +2253,48 @@ def import_pynvml():
+    methods.
+    """
+
+    def decorator(cls: Type[T]) -> Type[T]:
```
I think you can also remove this `decorator` function, and let the `warn_for_unimplemented_methods` directly accept the `cls` argument.

AoyuQC committed:
check NotImplementedError directly

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

AoyuQC committed:
remove decorator inside warn_for func

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

youkaichao commented on vllm/utils.py:
```diff
@@ -2253,3 +2253,39 @@ def import_pynvml():
+                base_method_name = base_method.__func__
+            else:
+                continue
+            class_method_name = getattr(cls, attr_name, False)
```
it seems it is still not checking `NotImplementedError`?

AoyuQC replied:
whoops, my bad, please check updated commit

youkaichao committed:
Merge branch 'main' into feature/v1-worker-base

AoyuQC committed:
add NotImplementedError check

Signed-off-by: Aoyu <aoyuzhan@amazon.com>

youkaichao committed:
use getsource

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'feature/v1-worker-base' of github.com:AoyuQC/vllm into feature/v1-worker-base

youkaichao committed:
catch error

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao reviewed (approved):
thanks for the simplification!

youkaichao committed:
fix spec decode

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
checked that all failures are caused by hf connection error. merging
